File: R/functions/text_tools.R
## Functions
## title_to_snake_case
## txt_strip
## substr_right


#' Converts "Column Title" to column_title
#'
#' @export
title_to_snake_case <- function(text) {
    return(tolower(
        paste(
            unlist(strsplit(text, '[ ]')), collapse='_')
        )
    )
}


#' Removes special characters from beginning and end of a string
#'
#' @export
txt_strip <- function(x, chars=' ') {
    chars <- unique(strsplit(chars, '')[[1]])

    for (char in chars) {

        # special characters
        if (char %in% c('(', ')')) {
            char <- paste0('\\', char)
        }

        x <- gsub(paste0('.*^', char, '+'), '', x)
        x <- gsub(paste0('+', char, '$'), '', x)
    }
    return(x)
}


#' see: https://stackoverflow.com/questions/7963898/extracting-the-last-n-characters-from-a-string-in-r
#'
#' @export
substr_right <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}

File: python/tools/text_tools.py
import re


def title_to_snake_case(text: str) -> str:
    """Convert 'Column Title' to 'column_title'."""
    return '_'.join(text.lower().split())


def substr_right(x: str, n: int) -> str:
    """Extract last n characters from a string."""
    return x[-n:] if len(x) >= n else x


def txt_strip(x: str, chars: str = ' ') -> str:
    """Remove leading and trailing characters."""
    char_list = list(set(chars))

    for char in char_list:
        # escape special regex characters
        if char in '()[]{}.*+?^$|\\':
            char = '\\' + char

        x = re.sub(f'^{char}+', '', x)
        x = re.sub(f'{char}+$', '', x)

    return x
